### PR TITLE
HPCC-23984 Implement correct logic in CEspConfig::ensureSDSSessionDomains()

### PR DESCRIPTION
**Pull request by wangkx**
1. Rename the ensureSDSSessionDomains() to ensureSDSSession();
2. Inside the ensureSDSSession(), the code for reading Session
    Domains settings is moved into new readSessionDomainsSetting().
    It is called only esp is starting;
3. The ensureSDSSession() is called in 2 scenarios: 1. after
    the readSessionDomainsSetting() is called and if esp is not detached
    from dali; 2. when esp is attached to dali if the ensureSDSSession()
    has not being called;
4. In the existing code, the EspHttpBinding::setSDSSession() is
    used to ensure that /Sessions/Process has an Application branch
    for the esp binding. It is called when EspHttpBinding is created.
    That does not work if esp is not attached to dali when the
    EspHttpBinding is created. In this PR, the related code is moved
    into the ensureSDSSession() so that the code can also be called
    when esp is attached to dali as needed.
    
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@jakesmith please review the second commit.

**Comment by wangkx** [comment]
I am going to rebase the code since https://github.com/hpcc-systems/HPCC-Platform/pull/13699 has been merged.

**Comment by wangkx** [comment]
@jakesmith the code has been rebased to the latest 7.8.x. Only one commit left. It is a part of the old commit for https://track.hpccsystems.com/browse/HPCC-23880. Please review.

File: esp/platform/espcfg.cpp [hunk 1]
```diff
@@ -220,35 +218,65 @@ void CEspConfig::ensureSDSSessionDomains()
...
-void CEspConfig::ensureESPSessionInTree(IPropertyTree* sessionRoot, const char* procName)
+void CEspConfig::ensureSDSSessionApplications(IPropertyTree* espSession)
 {
-    VStringBuffer xpath("%s[@name=\"%s\"]", PathSessionProcess, procName);
-    IPropertyTree* procSessionTree = sessionRoot->queryBranch(xpath.str());
-    if (!procSessionTree)
+    std::list<int> bindingPorts;
+
+    list<binding_cfg*>::iterator iter = m_bindings.begin();
+    while (iter!=m_bindings.end())
```
Comment (jakesmith) [inline esp/platform/espcfg.cpp:257]: would be simpler to use a for loop auto iterator, i.e.:
```
for (const auto pbfg: m_bindings)
```

**Comment by jakesmith** [comment]
@wangkx - looks good, 2 trivial comments.

**Comment by wangkx** [comment]
@jakesmith please review my changes.

**Comment by jakesmith** [comment]
@wangkx - looks good, please squash.

**Comment by wangkx** [comment]
@jakesmith  I just squashed.

**Comment by jakesmith** [comment]
@richardkchapman - ready to merge.

**Pull request merged**